gel/protocol/protocol.pyx

```
@@ -553,6 +645,137 @@ cdef class SansIOProtocol:
+                                self.buffer.discard_message()
+                    else:
+                        self.buffer.discard_message()
+
+                elif mtype == COMMAND_COMPLETE_MSG:
+                    self.parse_command_complete_message()
+                    results.append(result)
+                    result = []
+                    if ctx is ctxs[-1]:
+                        ctx = None
+                    else:
+                        ctx = next(it)
+
+                elif mtype == ERROR_RESPONSE_MSG:
+                    exc = self.parse_error_message()
+                    exc._query = ctx.query
+                    if exc.get_code() == parameter_type_mismatch_code:
+                        if not isinstance(ctx.in_dc, NullCodec):
+                            buf = WriteBuffer.new()
+                            try:
+                                self.encode_args(
+                                    ctx.in_dc, buf, ctx.args, ctx.kwargs
+                                )
+                            except errors.QueryArgumentError as ex:
+                                exc = ex
+                            else:
+                                # Now we know for sure that the new codec can
+                                # encode the same arguments, let's hint the
+                                # caller to retry as a special case.
+                                exc.tags = exc.tags.union(
+                                    {errors.SHOULD_RETRY}
+                                )
+
+                                # If we may retry this batch, invalidate codec
+                                # cache for all queries except the current one
+                                # because we just updated its input codec
+                                for c in ctxs:
+                                    if c is not ctx:
+                                        c.invalidate_cache()
```

We need to exit the loop or re-raise? Not sure I see what happens next after invalidation here